File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.go
func smallestDivisor(nums []int, threshold int) int {
    isPossible := func(mid int) bool {
        total := 0
        for _, num := range nums {
            total += int(math.Ceil(float64(num) / float64(mid)))
        }
        return total <= threshold
    }
    
    left, right := 1, getMax(nums)
    
    for left <= right {
        mid := left + (right - left) / 2
        
        if isPossible(mid) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    
    return left
}


func getMax(nums []int) int {
    max := math.MinInt32
    for _, num := range nums {
        if num > max {
            max = num
        }
    }
    return max
}

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.py
class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        def isPossible(mid, nums, threshold):
            total = 0
            for num in nums:
                total += -(-num // mid)

            return total <= threshold

        left, right = 1, max(nums)

        while left <= right:
            mid = left + (right - left) // 2

            if isPossible(mid, nums, threshold):
                right = mid - 1

            else:
                left = mid + 1

        return left
